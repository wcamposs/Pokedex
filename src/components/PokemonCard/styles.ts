// libraries
import { StyleSheet } from 'react-native';

// js
import colors from '../../utils/colors';

const styles = StyleSheet.create({
    container: (backgroundColor: String) => ({
        backgroundColor,

        borderWidth: 1,
        borderColor: backgroundColor,
        borderRadius: 16,
        marginVertical: 8,
        marginHorizontal: 12,
        overflow: 'hidden',
        alignItems: 'flex-start',

        //card shadow
        shadowColor: colors.black,
        shadowOffset: {
            width: 0,
            height: 4,
        },
        shadowOpacity: 0.32,
        shadowRadius: 5.46,
        elevation: 7,
    }),
    containerRow: {
        flexDirection: 'row',
    },
    imageContainer: {
        marginLeft: 20,
        maxHeight: 100,
        maxWidth: 100,
    },
    image: {
        width: 100,
        height: 100,
    },
    infoContainer: {
        marginLeft: 14,
        alignItems: 'center',
        justifyContent: 'center',
    },
    name: {
        color: colors.white,
        fontSize: 20,
        fontWeight: 'bold',
    },
    typesContainerRow: {
        flexDirection: 'row',
        marginTop: 12,
    },
    idContainer: {
        flex: 1,
        alignItems: 'flex-end',
        justifyContent: 'flex-end',
        marginHorizontal: 12,
        marginVertical: 4,
    },
    idText: {
        color: colors.darkGrey,
        fontSize: 24,
        fontWeight: 'bold',
    },
})

export default styles;