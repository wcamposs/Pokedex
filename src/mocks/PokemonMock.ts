const PokemonMock = {
	name: 'ivysaur',
	url: 'https://pokeapi.co/api/v2/pokemon/2/',
	id: '2',
	spriteUrl:
		'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
	types: [
		{
			slot: 1,
			type: {
				name: 'grass',
				url: 'https://pokeapi.co/api/v2/type/12/',
			},
		},
		{
			slot: 2,
			type: {
				name: 'poison',
				url: 'https://pokeapi.co/api/v2/type/4/',
			},
		},
	],
	abilities: [
		{
			ability: {
				name: 'overgrow',
				url: 'https://pokeapi.co/api/v2/ability/65/',
			},
			is_hidden: false,
			slot: 1,
		},
		{
			ability: {
				name: 'chlorophyll',
				url: 'https://pokeapi.co/api/v2/ability/34/',
			},
			is_hidden: true,
			slot: 3,
		},
	],
	height: 10,
	weight: 130,
	stats: [
		{
			base_stat: 60,
			effort: 0,
			stat: {
				name: 'hp',
				url: 'https://pokeapi.co/api/v2/stat/1/',
			},
		},
		{
			base_stat: 62,
			effort: 0,
			stat: {
				name: 'attack',
				url: 'https://pokeapi.co/api/v2/stat/2/',
			},
		},
		{
			base_stat: 63,
			effort: 0,
			stat: {
				name: 'defense',
				url: 'https://pokeapi.co/api/v2/stat/3/',
			},
		},
		{
			base_stat: 80,
			effort: 1,
			stat: {
				name: 'special-attack',
				url: 'https://pokeapi.co/api/v2/stat/4/',
			},
		},
		{
			base_stat: 80,
			effort: 1,
			stat: {
				name: 'special-defense',
				url: 'https://pokeapi.co/api/v2/stat/5/',
			},
		},
		{
			base_stat: 60,
			effort: 0,
			stat: {
				name: 'speed',
				url: 'https://pokeapi.co/api/v2/stat/6/',
			},
		},
	],
	evolutionChain: {
		evolution_details: [],
		evolves_to: [
			{
				evolution_details: [
					{
						gender: null,
						held_item: null,
						item: null,
						known_move: null,
						known_move_type: null,
						location: null,
						min_affection: null,
						min_beauty: null,
						min_happiness: null,
						min_level: 16,
						needs_overworld_rain: false,
						party_species: null,
						party_type: null,
						relative_physical_stats: null,
						time_of_day: '',
						trade_species: null,
						trigger: {
							name: 'level-up',
							url: 'https://pokeapi.co/api/v2/evolution-trigger/1/',
						},
						turn_upside_down: false,
					},
				],
				evolves_to: [
					{
						evolution_details: [
							{
								gender: null,
								held_item: null,
								item: null,
								known_move: null,
								known_move_type: null,
								location: null,
								min_affection: null,
								min_beauty: null,
								min_happiness: null,
								min_level: 32,
								needs_overworld_rain: false,
								party_species: null,
								party_type: null,
								relative_physical_stats: null,
								time_of_day: '',
								trade_species: null,
								trigger: {
									name: 'level-up',
									url: 'https://pokeapi.co/api/v2/evolution-trigger/1/',
								},
								turn_upside_down: false,
							},
						],
						evolves_to: [],
						is_baby: false,
						species: {
							name: 'venusaur',
							url: 'https://pokeapi.co/api/v2/pokemon-species/3/',
							spriteUrl:
								'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png',
						},
					},
				],
				is_baby: false,
				species: {
					name: 'ivysaur',
					url: 'https://pokeapi.co/api/v2/pokemon-species/2/',
					spriteUrl:
						'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png',
				},
			},
		],
		is_baby: false,
		species: {
			name: 'bulbasaur',
			url: 'https://pokeapi.co/api/v2/pokemon-species/1/',
			spriteUrl:
				'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
		},
	},
	growthRate: {
		name: 'medium-slow',
		url: 'https://pokeapi.co/api/v2/growth-rate/4/',
	},
	habitat: {
		name: 'grassland',
		url: 'https://pokeapi.co/api/v2/pokemon-habitat/3/',
	},
	legendary: false,
	mythical: false,
};

export default PokemonMock;
