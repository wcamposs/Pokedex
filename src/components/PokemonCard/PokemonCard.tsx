// libraries
import React from 'react';
import { Image, Text, TouchableOpacity, View } from 'react-native';

// js
import colors from '../../utils/colors';
import styles from './styles';
import backgroundColors from '../../utils/backgroundColors';

// components
import Chip from '../Chip/Chip';

// interfaces
import { PokemonItemProps } from '../../utils/interfaces';

const PokemonCard: React.FC<PokemonItemProps> = ({ pokemon, navigation }) => {
	const { id, name, spriteUrl, types } = pokemon;

	const typeName = types[0].type.name;
	const backgroundColor = backgroundColors[typeName] || colors.black;

	function handleNavigateToPokemonDetail() {
		navigation.navigate('PokemonDetails', { pokemon });
	}

	return (
		<TouchableOpacity
			style={styles.container(backgroundColor)}
			onPress={handleNavigateToPokemonDetail}>
			<View style={styles.containerRow}>
				<View style={styles.imageContainer}>
					<Image
						source={{ uri: spriteUrl }}
						style={styles.image}
						resizeMode="contain"
					/>
				</View>
				<View style={styles.infoContainer}>
					<View>
						<Text style={styles.name}>{name.toUpperCase()}</Text>

						<View style={styles.typesContainerRow}>
							{types.map((pokemonType: any) => {
								return (
									<React.Fragment key={pokemonType.type.name}>
										<Chip name={pokemonType.type.name} />
									</React.Fragment>
								);
							})}
						</View>
					</View>
				</View>

				<View style={styles.idContainer}>
					<Text style={styles.idText}>{`#${id}`}</Text>
				</View>
			</View>
		</TouchableOpacity>
	);
};

export default PokemonCard;
