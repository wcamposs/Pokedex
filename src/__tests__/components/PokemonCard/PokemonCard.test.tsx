import React from 'react';
import renderer from 'react-test-renderer';

import PokemonMock from '../../../mocks/PokemonMock';

import PokemonCard from '../../../components/PokemonCard/PokemonCard';

jest.useFakeTimers();

describe('<PokemonCard />', () => {
	it('renders correctly with an pokemon', () => {
		const navigation = jest.fn();

		const tree = renderer
			.create(
				<PokemonCard
					key={PokemonMock.name}
					pokemon={PokemonMock}
					navigation={navigation}
				/>,
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
