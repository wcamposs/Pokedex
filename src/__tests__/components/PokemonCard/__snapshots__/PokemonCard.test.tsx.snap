// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonCard /> renders correctly with an pokemon 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "flex-start",
      "backgroundColor": "#a3ccae",
      "borderColor": "#a3ccae",
      "borderRadius": 16,
      "borderWidth": 1,
      "elevation": 7,
      "marginHorizontal": 12,
      "marginVertical": 8,
      "opacity": 1,
      "overflow": "hidden",
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 4,
        "width": 0,
      },
      "shadowOpacity": 0.32,
      "shadowRadius": 5.46,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "marginLeft": 20,
          "maxHeight": 100,
          "maxWidth": 100,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png",
          }
        }
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": 14,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#FFF",
              "fontSize": 20,
              "fontWeight": "bold",
            }
          }
        >
          IVYSAUR
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "marginTop": 12,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 30,
                  "height": 20,
                  "justifyContent": "center",
                  "marginHorizontal": 4,
                  "maxHeight": 20,
                  "maxWidth": 100,
                  "textAlign": "center",
                  "width": 68,
                },
                Object {
                  "backgroundColor": "#00bf30",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 10,
                  "fontWeight": "bold",
                }
              }
            >
              GRASS
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 30,
                  "height": 20,
                  "justifyContent": "center",
                  "marginHorizontal": 4,
                  "maxHeight": 20,
                  "maxWidth": 100,
                  "textAlign": "center",
                  "width": 68,
                },
                Object {
                  "backgroundColor": "#b216f0",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 10,
                  "fontWeight": "bold",
                }
              }
            >
              POISON
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
          "justifyContent": "flex-end",
          "marginHorizontal": 12,
          "marginVertical": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#f5f5f5",
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
      >
        #2
      </Text>
    </View>
  </View>
</View>
`;
